src/main/scala/akka/contrib/process/BlockingProcess.scala

```diff
@@ -97,77 +97,59 @@ class BlockingProcess(
   override val supervisorStrategy: SupervisorStrategy =
     SupervisorStrategy.stoppingStrategy
 
-  private val process = startProcess()
-
-  // This shutdown hook is required because the stdio actors will be blocked on a read. The only
-  // reliable means of unblocking them is to destroy the thing they're listening to.
-  context.actorOf(ShutdownHook.props(() => process.destroy()), "process-destroyer")
+  // Using the constructor parameters here avoids having them needlessly retained
+  private val process: JavaProcess = {
+    import JavaConverters._
+    val preparedCommand = if (Helpers.isWindows) command map winQuote else command
+    val pb = new JavaProcessBuilder(preparedCommand.asJava)
+    pb.environment().putAll(environment.asJava)
+    pb.directory(directory)
+    pb.start()
+  }
 
-  private val stdin =
-    context.actorOf(OutputStreamSubscriber.props(process.getOutputStream), "stdin")
+  override def preStart(): Unit = {
+    val destroyer =
+      context.watch(context.actorOf(ProcessDestroyer.props(process), "process-destroyer"))
 
-  private val stdout =
-    context.watch(context.actorOf(InputStreamPublisher.props(process.getInputStream, stdioTimeout), "stdout"))
+    val stdin =
+      context.actorOf(OutputStreamSubscriber.props(process.getOutputStream), "stdin")
 
-  private val stderr =
-    context.watch(context.actorOf(InputStreamPublisher.props(process.getErrorStream, stdioTimeout), "stderr"))
+    val stdout =
+      context.watch(context.actorOf(InputStreamPublisher.props(process.getInputStream, stdioTimeout), "stdout"))
 
-  private var nrOfPublishers = 2
+    val stderr =
+      context.watch(context.actorOf(InputStreamPublisher.props(process.getErrorStream, stdioTimeout), "stderr"))
 
-  override def preStart(): Unit =
     receiver ! Started(ActorSubscriber(stdin), ActorPublisher(stdout), ActorPublisher(stderr))
+  }
 
   override def receive = {
     case Destroy =>
       log.debug("Received request to destroy the process.")
-      process.destroy()
-
-    case Terminated(publisher @ (`stdout` | `stderr`)) =>
-      log.debug("Publisher {} has terminated", publisher.path.name)
-      nrOfPublishers -= 1
-      if (nrOfPublishers == 0) {
-        val exitValue =
-          blocking {
-            process.waitFor()
-            process.exitValue()
-          }
-        receiver ! Exited(exitValue)
-        context.stop(self)
-      }
+      context.stop(self)
+    case Terminated(ref) =>
+      log.debug("Child {} was unexpectedly stopped, shutting down process", ref.path)
+      context.stop(self)
   }
 
-  private def startProcess(): JavaProcess = {
-    import JavaConverters._
-    val pb = new JavaProcessBuilder(prepareCommand(command).asJava)
-    pb.environment().putAll(environment.asJava)
-    pb.directory(directory)
-    pb.start()
+  override def postStop(): Unit = {
```

Comment:
Why not then put the `postStop` logic here in the process destroyer? We can then move the `process` val within the `preStart` and be assured that the process destroyer is the only thread that has access to it. WDYT?